File: clojure/test/nonamedb/test/storage/engines/memory_engine_spec.clj
(ns nonamedb.test.storage.engines.memory-engine-spec
  (:require
    [clojure.test :refer :all]
    [nonamedb.storage.engines.memory-engine :refer [get-from put-into create-engine]]))

(def test-key "some key")
(def test-value (byte-array (map byte "some value")))
(def updated-test-value (byte-array (map byte "some updated value")))

(deftest should-fail-to-retrieve-missing-data
  (let [engine (create-engine)]
    (is (nil? (get-from engine test-key)))
    (is (nil? (get-from engine "missing-data")))))

(deftest should-successfully-retrieve-data
  (let [engine (create-engine)]
    (is (put-into engine test-key test-value))
    (is (= (get-from engine test-key) test-value))))

(deftest should-successfully-retrieve-updated-data
  (let [engine (create-engine)]
    (is (put-into engine test-key test-value))
    (is (= (get-from engine test-key) test-value))
    (is (put-into engine test-key updated-test-value))
    (is (= (get-from engine test-key) updated-test-value))))

(deftest should-fail-to-retrieve-removed-data
  (let [engine (create-engine)]
    (is (put-into engine test-key test-value))
    (is (= (get-from engine test-key) test-value))
    (is (put-into engine test-key (byte-array [])))
    (is (nil? (get-from engine test-key)))))

File: scalafp/src/test/scala/nonamedb/test/specs/unit/storage/engines/MemoryEngineSpec.scala
package nonamedb.test.specs.unit.storage.engines

import cats.effect._
import nonamedb.storage.engines.memory.MemoryEngine
import nonamedb.test.specs.unit.UnitSpec

class MemoryEngineSpec extends UnitSpec {
  "A MemoryEngine" should "fail to retrieve missing data" in {
    for {
      engine <- MemoryEngine[IO]()
      result <- engine.get(testKey)
    } yield {
      result should be(None)
    }
  }

  it should "successfully add and retrieve data" in {
    for {
      engine <- MemoryEngine[IO]()
      _ <- engine.put(testKey, testValue)
      result <- engine.get(testKey)
    } yield {
      result should be(Some(testValue))
    }
  }

  it should "successfully update data" in {
    for {
      engine <- MemoryEngine[IO]()
      _ <- engine.put(testKey, testValue)
      initialResult <- engine.get(testKey)
      _ <- engine.put(testKey, updatedTestValue)
      updatedResult <- engine.get(testKey)
    } yield {
      initialResult should be(Some(testValue))
      updatedResult should be(Some(updatedTestValue))
    }
  }

  it should "successfully remove data" in {
    for {
      engine <- MemoryEngine[IO]()
      _ <- engine.put(testKey, testValue)
      initialResult <- engine.get(testKey)
      _ <- engine.put(testKey, Array.emptyByteArray)
      updatedResult <- engine.get(testKey)
    } yield {
      initialResult should be(Some(testValue))
      updatedResult should be(None)
    }
  }

  private val testKey = "some key"
  private val testValue = "some value".getBytes
  private val updatedTestValue = "some updated value".getBytes
}
